app/lib/oidc_client.rb

```diff
@@ -0,0 +1,121 @@
+require "openid_connect"
+
+class OidcClient
+  class OAuthFailure < RuntimeError; end
+
+  attr_reader :client_id,
+              :provider_uri
+
+  delegate :authorization_endpoint,
+           :token_endpoint,
+           :userinfo_endpoint,
+           :end_session_endpoint,
+           to: :discover
+
+  def initialize(provider_uri, client_id, secret)
+    @provider_uri = provider_uri
+    @client_id = client_id
+    @secret = secret
+  end
+
+  def auth_uri(return_path: nil)
+    nonce = SecureRandom.hex(16)
+    state = "#{nonce}:#{return_path}"
+
+    {
+      uri: client.authorization_uri(
+        scope: scopes,
+        state: state,
+        nonce: nonce,
+      ),
+      state: state,
+    }
+  end
+
+  def redirect_uri
+    host = ENV["GOVUK_WEBSITE_ROOT"] || "http://#{ENV['VIRTUAL_HOST']}"
```

Comment:
Should `VIRTUAL_HOST` come first?  In govuk-docker both env vars are set but it won't work as a redirect URI because there's no router.

Reply:
It probably should.  I don't have the `GOVUK_WEBSITE_ROOT` set, which is probably why I haven't seen an issue.